### PR TITLE
Replace gogoprotobuf with golang protobuf

### PR DESCRIPTION
**Pull request by hda2**
## Description

This change replaces gogoprotobuf with golang protobuf for Calico components. Gogoprotobuf is [deprecated](https://x.com/awalterschulze/status/1584553056100057088) since Oct 2022 and is not updated. Replacing gogoprotobuf with golang protobuf can bring in new protobuf features/fixes and reduce outdated package dependencies.

To work with golang protobuf, we need (1) a [calico/go-build change](https://github.com/projectcalico/go-build/pull/574) to include protoc in the toolchain image and (2) workaround an issue that protobuf generated types are not comparable. 

## Related issues/PRs

Require https://github.com/projectcalico/go-build/pull/574 to be merged and pick to release branches first.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: app-policy/checker/match_test.go [hunk 1]
```diff
@@ -351,10 +352,12 @@ func TestMatchRuleNamespaceSelectors(t *testing.T) {
 	}}
 
 	store := policystore.NewPolicyStore()
-	id := proto.NamespaceID{Name: "src"}
-	store.NamespaceByID[id] = &proto.NamespaceUpdate{Id: &id, Labels: map[string]string{"place": "src"}}
-	id = proto.NamespaceID{Name: "dst"}
-	store.NamespaceByID[id] = &proto.NamespaceUpdate{Id: &id, Labels: map[string]string{"place": "dst"}}
+	id := types.NamespaceID{Name: "src"}
+	pid := types.NamespaceIDToProto(id)
```
Comment (hjiawei) [inline app-policy/checker/match_test.go:356]: `pid` is a bit ambiguous. How about change it to `protoID`? This applies to other files in this PR.

File: app-policy/proto/healthz_grpc.pb.go
Comment (hjiawei) [inline app-policy/proto/healthz_grpc.pb.go:1]: This file is new. Do we need it?

File: cni-plugin/pkg/dataplane/grpc/proto/cnibackend_grpc.pb.go
Comment (hjiawei) [inline cni-plugin/pkg/dataplane/grpc/proto/cnibackend_grpc.pb.go:1]: This file is new. Do we need it?

File: felix/proto/felixbackend_grpc.pb.go
Comment (hjiawei) [inline felix/proto/felixbackend_grpc.pb.go:1]: This is a new file. Do we need it?

File: felix/types/host_endpoint_id.go
Comment (hjiawei) [inline felix/types/host_endpoint_id.go:1]: Prefer `hostendpoint_id.go` filename to align others. Same for the test file.

File: lib.Makefile [hunk 2]
```diff
@@ -135,7 +135,7 @@ endif
 # For building, we use the go-build image for the *host* architecture, even if the target is different
 # the one for the host should contain all the necessary cross-compilation tools
 # we do not need to use the arch since go-build:v0.15 now is multi-arch manifest
-GO_BUILD_IMAGE ?= calico/go-build
+GO_BUILD_IMAGE ?= jha48/go-build
```
Comment (hjiawei) [inline lib.Makefile:138]: We will revert this change once https://github.com/projectcalico/go-build/pull/574 is merged.

File: metadata.mk [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #################################################################################################
 
 # The version of github.com/projectcalico/go-build to use.
-GO_BUILD_VER=v0.91
+GO_BUILD_VER=protoc
```
Comment (hjiawei) [inline metadata.mk:6]: - [ ] We will revert this change once https://github.com/projectcalico/go-build/pull/574 is merged.

File: app-policy/checker/match_test.go (hunk 1 above)
Comment (hda2) [reply app-policy/checker/match_test.go]: changed all `pid` to `protoID`.

File: felix/proto/felixbackend_grpc.pb.go
Comment (hda2) [reply felix/proto/felixbackend_grpc.pb.go:1]: since we have defined the gRPC service:
https://github.com/projectcalico/calico/blob/41faf90b23bc827089fd1101e8f4adf5f265a756/felix/proto/felixbackend.proto#L25

We need this generated file.

File: app-policy/proto/healthz_grpc.pb.go
Comment (hda2) [reply app-policy/proto/healthz_grpc.pb.go:1]: Since we have defined the gRPC service:
https://github.com/projectcalico/calico/blob/41faf90b23bc827089fd1101e8f4adf5f265a756/app-policy/proto/healthz.proto#L6-L9

We need this generated file.

File: cni-plugin/pkg/dataplane/grpc/proto/cnibackend_grpc.pb.go
Comment (hda2) [reply cni-plugin/pkg/dataplane/grpc/proto/cnibackend_grpc.pb.go:1]: Since we have defined gRPC service:
https://github.com/projectcalico/calico/blob/41faf90b23bc827089fd1101e8f4adf5f265a756/cni-plugin/pkg/dataplane/grpc/proto/cnibackend.proto#L5-L8

We need this generated file.

File: felix/types/host_endpoint_id.go
Comment (hda2) [reply felix/types/host_endpoint_id.go:1]: file name was updated, thanks.

File: felix/calc/calc_graph_test.go [hunk 4]
```diff
@@ -70,7 +67,20 @@ var _ = DescribeTable("Calculation graph pass-through tests",
...
+		case *proto.IPAMPoolUpdate:
+			equal = googleproto.Equal(messageReceived.(*proto.IPAMPoolUpdate), expUpdate.(*proto.IPAMPoolUpdate))
+		case *proto.HostMetadataUpdate:
+			equal = googleproto.Equal(messageReceived.(*proto.HostMetadataUpdate), expUpdate.(*proto.HostMetadataUpdate))
+		case *proto.GlobalBGPConfigUpdate:
+			equal = googleproto.Equal(messageReceived.(*proto.GlobalBGPConfigUpdate), expUpdate.(*proto.GlobalBGPConfigUpdate))
+		case *proto.WireguardEndpointUpdate:
+			equal = googleproto.Equal(messageReceived.(*proto.WireguardEndpointUpdate), expUpdate.(*proto.WireguardEndpointUpdate))
+		case *proto.ServiceUpdate:
+			equal = googleproto.Equal(messageReceived.(*proto.ServiceUpdate), expUpdate.(*proto.ServiceUpdate))
```
Comment (hjiawei) [inline felix/calc/calc_graph_test.go:81]: Would it be more clear to duplicate `Expect(equal).To(BeTrue())` within each switch case? When the assertion fails, we immediately know which update message failed.

File: felix/calc/event_sequencer_test.go [hunk 5]
```diff
@@ -116,16 +116,14 @@ var _ = Describe("ParsedRulesToActivePolicyUpdate", func() {
...
-		numFields := protoPolicyValue.NumField()
-		for i := 0; i < numFields; i++ {
-			field := protoPolicyValue.Field(i)
-			fieldName := reflect.TypeOf(protoPolicy).Field(i).Name
-			Expect(field.Interface()).NotTo(Equal(reflect.Zero(field.Type()).Interface()),
-				fmt.Sprintf("Field %s in the protobuf struct was still at its zero value; "+
-					"missing from conversion or fully-loaded rule?", fieldName))
+		protoPolicy := protoUpdate.GetPolicy()
+		if googleproto.Equal(protoPolicy, &proto.Policy{}) {
+			return
```
Comment (hjiawei) [inline felix/calc/event_sequencer_test.go:121]: Should we do an `Expect` call here instead of `return`?

File: felix/calc/rule_convert.go [hunk 6]
```diff
@@ -63,7 +64,7 @@ func fillInRuleIDs(rules []*proto.Rule, ruleIDSeed string) {
 		// library.
 		// TODO(smc) Can we do better than hashing the protobuf?
 		rule.RuleId = ""
-		data, err := rule.Marshal()
+		data, err := googleproto.Marshal(rule)
 		if err != nil {
 			log.WithError(err).WithField("rule", rule).Panic("Failed to marshal rule")
```
Comment (hjiawei) [inline felix/calc/rule_convert.go:69]: May need to use `string(rule)`?

File: felix/calc/event_sequencer_test.go (hunk 5 above)
Comment (hda2) [reply felix/calc/event_sequencer_test.go]: this check is because I observed unit test panic caused by empty proto `protoPolicy.Namespace` reference. Since I updated all reference by using `Get`, this check should be safe to remove.

File: felix/calc/calc_graph_test.go (hunk 4 above)
Comment (hda2) [reply felix/calc/calc_graph_test.go]: updated.

File: felix/calc/rule_convert.go (hunk 6 above)
Comment (hda2) [reply felix/calc/rule_convert.go]: used `rule.String()`.

File: felix/types/workload_endpoint_id.go
Comment (hjiawei) [inline felix/types/workload_endpoint_id.go:1]: nit. `workloadendpoint` one word to align with other files.

File: felix/types/service_account_id.go
Comment (hjiawei) [inline felix/types/service_account_id.go:1]: nit. `serviceaccount` one word to align with other files.

File: felix/fv/policysync_test.go [hunk 7]
```diff
@@ -511,18 +515,18 @@ var _ = Context("_POL-SYNC_ _BPF-SAFE_ policy sync API tests", func() {
...
-										Id:     &saID,
-										Labels: map[string]string{"key.1": "value.1", "key_2": "value-2"},
-									},
-								}))
+								v := c.ServiceAccounts()
+								equal := googleproto.Equal(v[saID], &proto.ServiceAccountUpdate{
+									Id:     types.ServiceAccountIDToProto(saID),
+									Labels: map[string]string{"key.1": "value.1", "key_2": "value-2"},
+								})
+								Expect(equal).To(BeTrue())
```
Comment (hjiawei) [inline felix/fv/policysync_test.go:523]: This was an `Eventually` check but now changed to `Expect`. This test may need to check periodically and wait for it to pass (same below).

**Comment by hjiawei** [comment]
CI run for https://github.com/projectcalico/calico/pull/8949/commits/5fb90d5888c8629caf3affd84ec6514694be291b is [green](https://tigera.semaphoreci.com/workflows/a5a25e40-9593-47d4-8dec-2c498436cfb1?pipeline_id=59e1baa5-9f88-4e4e-8b73-682b6994e26b).

**Review by hjiawei — APPROVED** [review]
Thanks for you contribution @hda2.
